server/WFLY-10083_Hot_Class_Replacement.adoc

```diff
@@ -0,0 +1,132 @@
+enabled. This will not be enabled by default, but will require the user to start wildfly with a specific option
+(tentatively `--developer-mode`).
+
+In general there are three specific parts to this support:
+
+* Figuring out which classes need to be replaced, and providing them to the Agent
+* Updating the bytecode
+* Informing any frameworks that use these classes, so they can rebuild their internal data structures to reflect the
+changes
+
+===== Class change notification
+
+Hot replacement will only be supported for exploded deployments. There are two different ways that classes can be
+replaced:
+
+ * By replacing the classes inside the exploded deployment. The hot replacement support will install an Undertow handler
+ inside the deployment chain, that will check for changed classes when a request arrives (in much the same manner as the
+ old Seam 2 hot replacement filter).
+ * Using a debugger over the JDWP. In this case the class bytes will be updated on the filesystem
+ to keep the class file in sync with what is loaded into the VM. This means that if the
+ developer restarts the server or deployment the class change will be persistent.
+
+The main difficulty in supporting non-exploded deployments is that it is generally not possible to update the jar file
+with the new classes, so if the deployment is restarted it will have the old versions of the classes. It may be possible
+to address this in future via various strategies, such as queuing the updates and applying them on redeploy/restart.
+
+==== Replacing the Bytecode
+
+The Fakereplace agent will be responsible for replacing the bytecode of any changed classes. This operation is not
+guaranteed to succeed, as some changes are too invasive to be handled by the agent (e.g. changing the superclass or
+implemented interfaces).
+
+It may be desirable to do an automatic redeployment if the hot replacement fails, although this should probably be a
+configurable option.
+
+Note that only Fakereplace core will be consumed by WildFly. The existing integrations that are packaged in the dist
+jar are not going to be included, but will instead be provided by WildFly itself.
+
+==== Framework Support
+
+Support for Framework changes will be implemented in a combination of both Wildfly and the frameworks themselves. For
+some frameworks such as JAX-RS it should be possible to implement this in Wildfly directly, as all that should be required
+is a restart of the JAX-RS Servlet. For others such as Weld it will require potentially significant changes to the
+framework itself.
+
+Rather than forcing frameworks to consume Fakereplace as a dependency to use Fakereplace API's WildFly will act as the
+glue that connects Fakereplace and the frameworks. Essentually frameworks will provide their own APIs for hot replacement
+notification, and WildFly will bridge the gap between the frameworks and Fakereplace.
```

Comment:
What frameworks are targeted for this proposal?

Perhaps it would be better to have separate tracking for additional frameworks.